junit4-ant/src/main/java/com/carrotsearch/ant/tasks/junit4/slave/SlaveMain.java

```diff
@@ -245,12 +246,23 @@ public void testAssumptionFailure(Failure failure) {
         try {
           Runner runner = request.getRunner();
           methodFilter.apply(runner);
+
+          // New RunListener instances should be added per class and then removed from the RunNotifier
+          ArrayList<RunListener> runListenerInstances = UserDefinedRunListeners.generateInstances();
+
+          for(RunListener runListener : runListenerInstances) {
+            fNotifier.addListener(runListener);
+          }
   
           fNotifier.fireTestRunStarted(runner.getDescription());
           debug(debug, "Runner.run(" + clName + ")");
           runner.run(fNotifier);
           debug(debug, "Runner.done(" + clName + ")");
           fNotifier.fireTestRunFinished(result);
+
```

Comment:
Why do they have to be unregistered from the notifier? Just curious.

Reply:
I find it good practice to add RunListener instances at a class level and then remove them after the class is done executing.  This is helpful if there is any common data shared for the life of a RunListener instance at the class level.  We do have listeners that have common data for processing test metadata